### PR TITLE
Chore/remove old create collection code

### PR DESCRIPTION
**Pull request by mchacki**
### Scope & Purpose

*Cleanup Task: Removes now obsolete createCollectionCode.
It had to rewire some methods that are still in use.
Moved Parts of ClusterInfo into TestOnly code, some C++ tests rely on it, and cannot be trivially moved to official APIs as those tests actually create invalid collections (which does not matter for the test, as the collection is just a starting point for Index meta data tests.)
*

- [ ] :hankey: Bugfix
- [ ] :pizza: New feature
- [ ] :fire: Performance improvement
- [x] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.11: *(Please link PR)*
  - [ ] Backport for 3.10: *(Please link PR)*
  - [ ] Backport for 3.9: *(Please link PR)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [x] Enterprise PR: https://github.com/arangodb/enterprise/pull/1336
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

File: arangod/VocBase/Methods/UpgradeTasks.cpp [hunk 1]
```diff
@@ -317,40 +317,23 @@ Result createSystemStatisticsCollections(
     };
     std::vector<std::shared_ptr<VPackBuffer<uint8_t>>> buffers;
     Result res;
+    OperationOptions options(ExecContext::current());
```
Comment (jsteemann) [inline arangod/VocBase/Methods/UpgradeTasks.cpp:320]: Do we really want to execute UpgradeTasks with the current user's permissions? 
We need to create collections here, potentially even in the `_system` database.
Shouldn't we run the UpgradeTasks with root permissions, always?

File: arangod/VocBase/Methods/UpgradeTasks.cpp [hunk 2]
```diff
@@ -317,40 +317,23 @@ Result createSystemStatisticsCollections(
...
       if (res.fail()) {
         return res;
       }
-      // capture created collection vector
-      createdCollections.insert(std::end(createdCollections), std::begin(cols),
-                                std::end(cols));
+      TRI_ASSERT(col) << "Create system collection did not fail but also did "
+                         "not create a collection.";
+      createdCollections.emplace_back(std::move(col));
+      res = methods::Collections::lookup(vocbase, collection, col);
```
Comment (jsteemann) [inline arangod/VocBase/Methods/UpgradeTasks.cpp:336]: This is a use-after-move for `col`.

File: arangod/VocBase/Methods/UpgradeTasks.cpp [hunk 3]
```diff
@@ -359,27 +342,8 @@ Result createSystemStatisticsCollections(
...
-    return methods::Collections::create(
-        vocbase, operationOptions, {info}, true, true, true,
-        nullptr /*nullptr on purpose, distributeShardsLike is defined by
-                   createSystemCollectionProperties */
-        ,
-        cols, true /* allow system collection creation */);
-  }
-
-  return {TRI_ERROR_NO_ERROR};
+  OperationOptions operationOptions(ExecContext::current());
```
Comment (jsteemann) [inline arangod/VocBase/Methods/UpgradeTasks.cpp:345]: Does this somehow change how permissions are handled for creating the collections?

File: tests/Mocks/ClusterInfoMock.cpp [hunk 4]
```diff
@@ -0,0 +1,989 @@
...
+#include "Utils/Events.h"
+#include "VocBase/LogicalCollection.h"
+#include "Replication2/Methods.h"
+#include "Replication2/StateMachines/Document/DocumentStateMachine.h"
+
+using namespace arangodb;
+using namespace arangodb::cluster;
+
+namespace {
+inline arangodb::AgencyOperation IncreaseVersion() {
```
Comment (jsteemann) [inline tests/Mocks/ClusterInfoMock.cpp:48]: nit: I don't think that any of these functions need to be declared `inline`.
They have static linkage anyway.

File: tests/Mocks/ClusterInfoMock.cpp [hunk 5]
```diff
@@ -0,0 +1,989 @@
...
+  return arangodb::AgencyOperation{
+      "Plan/Version", arangodb::AgencySimpleOperationType::INCREMENT_OP};
+}
+
+inline std::string collectionPath(std::string_view dbName,
+                                  std::string_view collection) {
+  return "Plan/Collections/" + std::string{dbName} + "/" +
+         std::string{collection};
```
Comment (jsteemann) [inline tests/Mocks/ClusterInfoMock.cpp:56]: nit: use absl::StrCat(...)` for simpler and more efficient code.

File: tests/Mocks/ClusterInfoMock.cpp [hunk 6]
```diff
@@ -0,0 +1,989 @@
...
+          info, arangodb::StaticStrings::IsSmart, false)) {
+    TRI_ASSERT(info.hasKey(arangodb::StaticStrings::AttrIsBuilding));
+    TRI_ASSERT(info.get(arangodb::StaticStrings::AttrIsBuilding).isBool());
+    TRI_ASSERT(info.get(arangodb::StaticStrings::AttrIsBuilding).getBool() ==
+               true);
+  }
+#endif
+  arangodb::AgencyOperation op{collectionPath(dbName, collection),
+                               arangodb::AgencyValueOperationType::SET, info};
+  return op;
```
Comment (jsteemann) [inline tests/Mocks/ClusterInfoMock.cpp:73]: nit: these 2 statements could be fused into a single one.

File: tests/Mocks/ClusterInfoMock.cpp [hunk 7]
```diff
@@ -0,0 +1,989 @@
...
+}
+
+inline arangodb::AgencyPrecondition CreateCollectionOrderPrecondition(
+    std::string_view dbName, std::string_view collection, VPackSlice info) {
+  arangodb::AgencyPrecondition prec{collectionPath(dbName, collection),
+                                    arangodb::AgencyPrecondition::Type::VALUE,
+                                    info};
+  return prec;
```
Comment (jsteemann) [inline tests/Mocks/ClusterInfoMock.cpp:81]: nit: these 2 statements could be fused into a single one.

File: tests/Mocks/ClusterInfoMock.cpp [hunk 8]
```diff
@@ -0,0 +1,989 @@
...
+  // closure and the main thread executing this function. Note that it can
+  // happen that the callback is called only after we return from this
+  // function!
+  auto dbServerResult =
+      std::make_shared<std::atomic<std::optional<ErrorCode>>>(std::nullopt);
+  auto nrDone = std::make_shared<std::atomic<size_t>>(0);
+  auto errMsg = std::make_shared<std::string>();
+  auto cacheMutex = std::make_shared<std::mutex>();
+  auto cacheMutexOwner = std::make_shared<std::atomic<std::thread::id>>();
+  auto isCleaned = std::make_shared<bool>(false);
```
Comment (jsteemann) [inline tests/Mocks/ClusterInfoMock.cpp:333]: Note: these changes will conflict with an open PR from @MBkkt : https://github.com/arangodb/arangodb/pull/19503

File: tests/Mocks/ClusterInfoMock.cpp [hunk 9]
```diff
@@ -0,0 +1,989 @@
...
+      // We register a callback in the agency.
+      // For some reason this scopeguard is executed (e.g. error case)
+      // While we are in this cleanup, and before a callback is removed from the
+      // agency. The callback is triggered by another thread. We have the
+      // following guarantees: a) cacheMutex|Owner are valid and locked by
+      // cleanup b) isCleaned is valid and now set to true c) the closure is
+      // owned by the callback d) info might be deleted, so we cannot use it. e)
+      // If the callback is ongoing during cleanup, the callback will
+      //    hold the Mutex and delay the cleanup.
+      RECURSIVE_MUTEX_LOCKER(*cacheMutex, *cacheMutexOwner);
```
Comment (jsteemann) [inline tests/Mocks/ClusterInfoMock.cpp:350]: same

File: arangod/VocBase/Methods/UpgradeTasks.cpp (hunk 2 above)
Comment (mchacki) [reply arangod/VocBase/Methods/UpgradeTasks.cpp:336]: 😲

File: arangod/VocBase/Methods/UpgradeTasks.cpp (hunk 1 above)
Comment (mchacki) [reply arangod/VocBase/Methods/UpgradeTasks.cpp:320]: Good point, this needs full access.
Adapted.

File: arangod/VocBase/Methods/UpgradeTasks.cpp (hunk 3 above)
Comment (mchacki) [reply arangod/VocBase/Methods/UpgradeTasks.cpp:345]: Yes change unintentional.
Fixed!

File: arangod/VocBase/Methods/UpgradeTasks.cpp (hunk 2 above)
Comment (mchacki) [reply arangod/VocBase/Methods/UpgradeTasks.cpp:336]: Fixed, the lookup was not necessary anyhow.

File: tests/Mocks/ClusterInfoMock.cpp (hunk 5 above)
Comment (mchacki) [reply tests/Mocks/ClusterInfoMock.cpp:56]: In general I agree this is just mock-code we need to keep around until we have a better way to write the tests. Hence I would like to keep it as is, there eis no need for performance optimization here.

File: tests/Mocks/ClusterInfoMock.cpp (hunk 8 above)
Comment (mchacki) [reply tests/Mocks/ClusterInfoMock.cpp:333]: Thanks for this 👍 Adapted the code to the new open PR.

**Review by jsteemann — APPROVED** [review]
LGTM

**Pull request merged**